crates/bevy_ecs/src/archetype.rs

```diff
@@ -0,0 +1,540 @@
+        &mut self.unique_components
+    }
+
+    #[inline]
+    pub fn components(&self) -> impl Iterator<Item = ComponentId> + '_ {
+        self.components.indices()
+    }
+
+    #[inline]
+    pub fn edges(&self) -> &Edges {
+        &self.edges
+    }
+
+    #[inline]
+    pub(crate) fn edges_mut(&mut self) -> &mut Edges {
+        &mut self.edges
+    }
+
+    #[inline]
+    pub fn entity_table_row(&self, index: usize) -> usize {
+        self.table_info.entity_rows[index]
+    }
+
+    #[inline]
+    pub fn set_entity_table_row(&mut self, index: usize, table_row: usize) {
+        self.table_info.entity_rows[index] = table_row;
+    }
+
+    /// # Safety
+    /// valid component values must be immediately written to the relevant storages
+    /// `table_row` must be valid
+    pub unsafe fn allocate(&mut self, entity: Entity, table_row: usize) -> EntityLocation {
+        self.entities.push(entity);
+        self.table_info.entity_rows.push(table_row);
+
+        EntityLocation {
+            archetype_id: self.id,
+            index: self.entities.len() - 1,
+        }
+    }
+
+    pub fn reserve(&mut self, additional: usize) {
+        self.entities.reserve(additional);
+        self.table_info.entity_rows.reserve(additional);
+    }
+
+    /// Removes the entity at `index` by swapping it out. Returns the table row the entity is stored in.
+    pub(crate) fn swap_remove(&mut self, index: usize) -> ArchetypeSwapRemoveResult {
+        let is_last = index == self.entities.len() - 1;
+        self.entities.swap_remove(index);
+        ArchetypeSwapRemoveResult {
+            swapped_entity: if is_last {
+                None
+            } else {
+                Some(self.entities[index])
+            },
+            table_row: self.table_info.entity_rows.swap_remove(index),
+        }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.entities.len()
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.entities.is_empty()
+    }
+
+    #[inline]
+    pub fn contains(&self, component_id: ComponentId) -> bool {
+        self.components.contains(component_id)
+    }
+
+    #[inline]
+    pub fn get_storage_type(&self, component_id: ComponentId) -> Option<StorageType> {
+        self.components
+            .get(component_id)
+            .map(|info| info.storage_type)
+    }
+
+    #[inline]
+    pub fn get_archetype_component_id(
+        &self,
+        component_id: ComponentId,
+    ) -> Option<ArchetypeComponentId> {
+        self.components
+            .get(component_id)
+            .map(|info| info.archetype_component_id)
+    }
+}
+
+/// A generational id that changes every time the set of archetypes changes
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct ArchetypeGeneration(usize);
+
+impl ArchetypeGeneration {
+    #[inline]
+    pub fn new(generation: usize) -> Self {
+        ArchetypeGeneration(generation)
+    }
+
+    #[inline]
+    pub fn value(&self) -> usize {
```

Comment:
```suggestion
    pub fn value(self) -> usize {
```
`ArchetypeGeneration` is `Copy`.